source/opt/local_access_chain_convert_pass.cpp

```diff
@@ -0,0 +1,366 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "local_access_chain_convert_pass.h"
+#include "iterator.h"
+
+/*
+#define SPV_FUNCTION_CALL_FUNCTION_ID 2
+#define SPV_FUNCTION_CALL_ARGUMENT_ID 3
+#define SPV_FUNCTION_PARAMETER_RESULT_ID 1
+#define SPV_STORE_POINTER_ID 0
+#define SPV_STORE_OBJECT_ID 1
+#define SPV_RETURN_VALUE_ID 0
+#define SPV_TYPE_POINTER_STORAGE_CLASS 1
+#define SPV_TYPE_POINTER_TYPE_ID 2
+*/
+
+static const int kSpvEntryPointFunctionId = 1;
+static const int kSpvStorePtrId = 0;
+static const int kSpvStoreValId = 1;
+static const int kSpvLoadPtrId = 0;
+static const int kSpvAccessChainPtrId = 0;
+static const int kSpvTypePointerStorageClass = 0;
+static const int kSpvTypePointerTypeId = 1;
+static const int kSpvConstantValue = 0;
+
+namespace spvtools {
+namespace opt {
+
+bool LocalAccessChainConvertPass::IsMathType(const ir::Instruction* typeInst) {
+  switch (typeInst->opcode()) {
+  case SpvOpTypeInt:
+  case SpvOpTypeFloat:
+  case SpvOpTypeBool:
+  case SpvOpTypeVector:
+  case SpvOpTypeMatrix:
+    return true;
+  default:
+    break;
+  }
+  return false;
+}
+
+bool LocalAccessChainConvertPass::IsTargetType(
+    const ir::Instruction* typeInst) {
+  if (IsMathType(typeInst))
+    return true;
+  if (typeInst->opcode() != SpvOpTypeStruct &&
+      typeInst->opcode() != SpvOpTypeArray)
+    return false;
+  int nonMathComp = 0;
+  typeInst->ForEachInId([&nonMathComp,this](const uint32_t* tid) {
+    ir::Instruction* compTypeInst =
+        def_use_mgr_->id_to_defs().find(*tid)->second;
+    // Ignore length operand in Array type
+    if (compTypeInst->opcode() == SpvOpConstant) return;
+    if (!IsMathType(compTypeInst)) ++nonMathComp;
+  });
+  return nonMathComp == 0;
+}
+
+ir::Instruction* LocalAccessChainConvertPass::GetPtr(
+    ir::Instruction* ip,
+    uint32_t* varId) {
+  const uint32_t ptrId = ip->opcode() == SpvOpStore ?
+      ip->GetSingleWordInOperand(kSpvStorePtrId) :
+      ip->GetSingleWordInOperand(kSpvLoadPtrId);
+  ir::Instruction* ptrInst =
+    def_use_mgr_->id_to_defs().find(ptrId)->second;
+  *varId = ptrInst->opcode() == SpvOpAccessChain ?
+    ptrInst->GetSingleWordInOperand(kSpvAccessChainPtrId) :
+    ptrId;
+  return ptrInst;
+}
+
+bool LocalAccessChainConvertPass::IsTargetVar(uint32_t varId) {
+  if (seen_non_target_vars_.find(varId) != seen_non_target_vars_.end())
+    return false;
+  if (seen_target_vars_.find(varId) != seen_target_vars_.end())
+    return true;
+  const ir::Instruction* varInst =
+    def_use_mgr_->id_to_defs().find(varId)->second;
+  assert(varInst->opcode() == SpvOpVariable);
```

Comment:
If I read the caller correctly, I think this code will assert out if you have:

%f = OpFunction...
 %var = OpVariable  ...
  %ptr_a = OpAccessChain %var %zero
  %ptr_b = OpAccessChain %ptr_a %zero
  %value = OpLoad %ptr_b
 
This will assert out because varInst is %ptr_a in this case.